Add proposal for adopting Kafka Connect health check endpoint (#139)
strimzi/proposals: 087-adopt-connect-health-endpoint.md

# Adopt the Kafka Connect health check endpoint

In a Kubernetes node, the kubelet component uses the configured liveness probe to know when to restart a container (i.e. deadlock).

> Instead, it uses the configured readiness probe to know when a container is ready to start accepting traffic (i.e. startup phase completed).

Reviewer: The `Instead` does not seem to fit into how the text flows? Also, the test seems a but confusing with regards to startup probes. Maybe you could just focus that it is used to verify if the pod is ready to accept traffic?

> A common pattern for liveness probes is to use the same HTTP endpoint as for readiness probes.

Reviewer: I don't think that is a common pattern ... but if you have nothing better, it is an option.

Author: Getting rid of this as not important in this context.

  "kafka_cluster_id": "_xIIeVIOQMKXimv_l96WtQ"
}
```

## Motivation

Using the `/` endpoint for Kafka Connect and Mirror Maker 2 health checks is a common approach, but it does not actually test for readiness, because these requests can be completed before the worker is started.
Instead, the `/health` endpoint waits for the worker startup phase to complete, which is made of internal topics creation if they do not exists, internal topics full read, and cluster join.

If the worker has not yet completed the startup phase, or it is unable to respond in time, the response will have a 5xx status code.
All Kafka Connect endpoints have a timeout of 90 seconds, with the exception of the `/health` endpoint where it is hardcoded to 10 seconds for faster unhealty worker detection.
Unlike the `/` endpoint, the `/health` endpoint response message includes error details that can help with troubleshooting.

Example output with `/health` endpoint:

```sh
$ kubectl exec my-cluster-kafka-0 -- curl -s http://my-connect-cluster-connect-api:8083/health | jq
{
  "status": "healthy",
  "message": "Worker has completed startup and is ready to handle requests."
}
```

## Proposal

> Strimzi will support the Kafka 3.8 branch for an extended period of time ([SIP-77](https://github.com/strimzi/proposals/blob/main/077-support-for-kafka-4.0.md#extended-support-for-kafka-39)) due to the upcoming major Kafka release.

Author: Removing the extended support part (my bad), I still think that it would be good to provide `/health` endpoint support as early as possible. The change to get rid of the 3.8 endpoint fallback logic is very simple, and there will be unit tests failing as a reminder. Wdyt?

> The `/health` endpoint will be used for Kafka Connect and Mirror Maker 2 components only when the following conditions will be true at the same time:

Reviewer: So, will it be used only for the readiness probe? Or for both readiness and liveness? ny changes tot he default timing of the probes?

Author: For both, added. Instead, the following line was already present:

> Default liveness and readiness probes configuration won't change.

> 1. `.spec.image` is null

Author: Ok, sounds good.

2. `.spec.version` is either null or doesn't start with "3.8"

When Kafka 3.8 branch will be out of support, we will be able to get rid of this logic, and always use the `/health` endpoint.
Unit tests related to this feature will fail when Kafka 3.8 support will be removed from the code base.

Default liveness and readiness probes configuration won't change.

## Affected/not affected projects

The only affected project is the Cluster Operator, in particular Kafka Connect and Mirror Maker 2 components.

## Compatibility

This change is fully backwards compatible as no further restrictions are created by the `/health` endpoint.

## Rejected alternatives

Wait until Kafka 3.8 is out of support in Strimzi and then switch from `/` to `/health`.

> It is important to adopt new Kafka features as early as possible, and this approach would add a big delay considering Kafka 3.8 extended support.

Author: Ok. See my other comment on this. Thanks.

Instead, the `/health` endpoint waits for the worker startup phase to complete, which is made of internal topics creation if they do not exists, internal topics full read, and cluster join.

If the worker has not yet completed the startup phase, or it is unable to respond in time, the response will have a 5xx status code.
All Kafka Connect endpoints have a timeout of 90 seconds, with the exception of the `/health` endpoint where it is hardcoded to 10 seconds for faster unhealty worker detection.
Unlike the `/` endpoint, the `/health` endpoint response message includes error details that can help with troubleshooting.

Example output with `/health` endpoint (probes only care about the HTTP response status code):

```sh
$ kubectl exec my-cluster-kafka-0 -- curl -s http://my-connect-cluster-connect-api:8083/health | jq
{
  "status": "healthy",
  "message": "Worker has completed startup and is ready to handle requests."
}
```

## Proposal

Strimzi currently supports both Kafka 3.8 and 3.9 releases.

> As the `/health` endpoint was introduced in Kafka 3.9, we will wait for Kafka 3.8 to go out of support in Strimzi before doing the switch.

Reviewer: Maybe you can specifically mention Strimzi 0.46 as that is where the bulk of the Kafka 4.0 and ZooKeeper removal changes will land.

A notable change will be added to the changelog to inform the users.

The configuration options for liveness and readiness probes won't change.

## Affected/not affected projects

The only affected project is the Cluster Operator, in particular Kafka Connect and MirrorMaker 2 components.

## Compatibility

This change is backwards compatible, and there should be no need to update Kafka Connect and MirrorMaker 2 probe configurations.

The following test results show that there isn't a significant difference in performance between the `/health` and `/` endpoints.
Note: pod ready time does not include the image pull time, and response time is computed as 95p over 200 requests with 10 seconds period.

| Endpoint | Pod ready time in seconds | Response time in ms |
|----------|---------------------------|---------------------|
| /        | 65                        | 3.0286              |
| /health  | 62                        | 3.7525              |

## Rejected alternatives

Switch to the `/health` endpoint while still supporting Kafka 3.8.

> This would allow to get the new feature early, but at the cost of additional complexity.

Reviewer 2: ```suggestion
This would allow the new feature to be adopted earlier, but with the cost of additional complexity.
```

# Adopt the Kafka Connect health check endpoint

> In a Kubernetes node, the kubelet component uses the configured liveness probe to know when to restart a container (i.e. deadlock).

Reviewer 3: ```suggestion
In a Kubernetes node, the kubelet component uses the configured liveness probe to know when to restart a container. 
With a suitably written application this can be used to restart a container if the applications process enters an illegal state (e.g. deadlock).
```

Author: Thanks Tom.

Unlike the `/` endpoint, the `/health` endpoint response message includes error details that can help with troubleshooting.

Example output with `/health` endpoint (probes only care about the HTTP response status code):

```sh
$ kubectl exec my-cluster-kafka-0 -- curl -s http://my-connect-cluster-connect-api:8083/health | jq
{
  "status": "healthy",
  "message": "Worker has completed startup and is ready to handle requests."
}
```

## Proposal

Strimzi supports both Kafka 3.8 and 3.9 branches.

The `/health` endpoint will be used for both readiness and liveness probes of Kafka Connect and Mirror Maker 2 components only when `.spec.version` is either null or doesn't start with "3.8".

When Kafka 3.8 branch will be out of support, we will be able to get rid of the 3.8 fallback logic, and always use the `/health` endpoint.
Unit tests related to this feature will fail when Kafka 3.8 support will be removed from the code base.

Default liveness and readiness probes configuration won't change.

## Affected/not affected projects

The only affected project is the Cluster Operator, in particular Kafka Connect and Mirror Maker 2 components.

> ## Compatibility

Author: > You explain above how the new health check is better ... but here, you say it does not matter. 

I'm not saying that. I'm just saying that performance is not a concern.

> The test you did below is nice, but I assume that it is a very simple and small cluster with the default configuration.

We can repeat that with a more realistic setup if that helps to gain confidence, but a feature gate seems to be an overkill to me.

> ## Compatibility

Reviewer: > We can repeat that with a more realistic setup if that helps to gain confidence

It is not about _realistic_ setup. It is about the extreme setups where it is most likely to break someone's setup.
